### DISCUSSION TITLE
TCA Dependencies

### DISCUSSION THREAD

**Discussion by katovBash** [pointfreeco/swift-composable-architecture#2218] [Q&A] [answered]
### Description

Hello!

I have discussion w/ developers in our team :)
Can u share your opinion on this topic please?

We have using dependencies with liveValue and testValue.
We have a lot of different dependencies and mocks to them. 

Would we save our mocks in the main target of the directory of our dependency and use it in testValue?

Or would we save all mocks in test target and change dependency implementation in our tests?

### Checklist

- [X] I have determined whether this bug is also reproducible in a vanilla SwiftUI project.
- [X] If possible, I've reproduced the issue using the `main` branch of this package.
- [X] This issue hasn't been addressed in an [existing GitHub issue](https://github.com/pointfreeco/swift-composable-architecture/issues) or [discussion](https://github.com/pointfreeco/swift-composable-architecture/discussions).

**Comment by mbrandonw** [comment]
Hi @katovBash, I am going to convert this to a discussion since it's not an issue or bug with the library.

**Comment by mbrandonw** [comment, marked as answer]
Hi @katovBash, in our view the most optimal implementation of the `testValue` of a dependency is the one that simply performs an `XCTFail` in every single endpoint, or uses the `unimplemented` helper to stub any kind of closure with one that performs `XCTFail` under the hood. This is what we do in our standups demo app:

https://github.com/pointfreeco/standups/blob/e130ec11cbd7de569256373fb177ae7cbdb470a0/Standups-StackBased/Standups/Dependencies/SpeechClient.swift#L72-L76

```swift
static let testValue = SpeechClient(
  authorizationStatus: unimplemented("SpeechClient.authorizationStatus", placeholder: .denied),
  requestAuthorization: unimplemented("SpeechClient.requestAuthorization", placeholder: .denied),
  startTask: unimplemented("SpeechClient.startTask")
)
```

…as well as every other demo and case study we have in our various repos.

This way you are immediately alerted to the moment a dependency starts being used in a feature because you will get a test failure in all the places you haven't dealt with that new behavior.

This also means you don't need to put any mocks in the module that holds the `TestDependencyKey` conformance. All of the mocks can continue living in your test target or test helper modules.

**Reply by katovBash** [reply]
Thanks a lot ❤️

**Reply by katovBash** [reply]
I have minor question about dependency, I'm trying to implement this approach but have some issue with it.

I have dependency like:

  static public var testValue: SmartDeviceServiceProtocol {
    unimplemented("\(Self.self).smartDeviceServiceMock")
  }

And when I'm testing it I change this dependency w/ mock, like:

    let reducer = Reducer()
      .dependency(
        \. smartDeviceService,
         smartDeviceServiceMock
      )

But I have exception in ide that I don't implement test value, can u please share your opinion, how can I avoid it?

**Reply by mbrandonw** [reply]
There's not enough information here to diagnose what is happening. Can you share more code?

**Reply by katovBash** [reply]
Thanks for your attention)
Also I have solved by changing .dependency to prepareDependencies in TestStore)
All works great!
